File: web/src/frontends/classic/app/page-routes.ts
import pageRouteManifest from '../../../../../internal/webui/page_routes.json'

export interface PageRouteEntry {
  readonly name: string
  readonly path: string
}

interface UnknownRecord {
  readonly [key: string]: unknown
}

const staticSegmentPattern = /^[A-Za-z0-9][A-Za-z0-9._~-]*$/
const parameterSegmentPattern = /^:[A-Za-z][A-Za-z0-9_]*$/
const manifestFields = new Set(['version', 'routes'])
const routeFields = new Set(['name', 'path'])

function isRecord(value: unknown): value is UnknownRecord {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function isSharedPathPattern(path: string): boolean {
  if (path === '/') return true
  if (!path.startsWith('/') || path.endsWith('/')) return false

  return path
    .slice(1)
    .split('/')
    .every(
      (segment) =>
        parameterSegmentPattern.test(segment) ||
        (segment !== '.' && segment !== '..' && staticSegmentPattern.test(segment)),
    )
}

function unknownField(record: UnknownRecord, allowed: ReadonlySet<string>): string | undefined {
  return Object.keys(record).find((field) => !allowed.has(field))
}

function pageRouteShape(path: string): string {
  return path
    .split('/')
    .map((segment) => (parameterSegmentPattern.test(segment) ? ':' : segment))
    .join('/')
}

export function parsePageRouteManifest(value: unknown): readonly PageRouteEntry[] {
  if (!isRecord(value)) {
    throw new Error('Invalid page route manifest: must be an object')
  }
  const unexpectedManifestField = unknownField(value, manifestFields)
  if (unexpectedManifestField !== undefined) {
    throw new Error(`Invalid page route manifest: unknown field "${unexpectedManifestField}"`)
  }
  if (value.version !== 1) {
    throw new Error('Invalid page route manifest: version must be 1')
  }
  if (!Array.isArray(value.routes)) {
    throw new Error('Invalid page route manifest: routes must be an array')
  }
  if (value.routes.length === 0) {
    throw new Error('Invalid page route manifest: routes must not be empty')
  }

  const names = new Set<string>()
  const paths = new Set<string>()
  const shapes = new Set<string>()
  const entries = value.routes.map((route, index) => {
    if (!isRecord(route)) {
      throw new Error(`Invalid page route manifest: route at index ${index} must be an object`)
    }
    const unexpectedRouteField = unknownField(route, routeFields)
    if (unexpectedRouteField !== undefined) {
      throw new Error(
        `Invalid page route manifest: route at index ${index} has unknown field "${unexpectedRouteField}"`,
      )
    }
    if (
      typeof route.name !== 'string' ||
      route.name.trim() === '' ||
      route.name.trim() !== route.name
    ) {
      throw new Error(
        `Invalid page route manifest: route at index ${index} must have a non-empty name`,
      )
    }
    if (typeof route.path !== 'string' || route.path.trim() === '') {
      throw new Error(
        `Invalid page route manifest: route at index ${index} must have a non-empty path`,
      )
    }
    if (!isSharedPathPattern(route.path)) {
      throw new Error(
        `Invalid page route manifest: route "${route.name}" uses an unsupported shared path pattern`,
      )
    }
    if (names.has(route.name)) {
      throw new Error(`Invalid page route manifest: duplicate route name "${route.name}"`)
    }
    if (paths.has(route.path)) {
      throw new Error(`Invalid page route manifest: duplicate route path "${route.path}"`)
    }
    const shape = pageRouteShape(route.path)
    if (shapes.has(shape)) {
      throw new Error(`Invalid page route manifest: duplicate route shape "${shape}"`)
    }

    names.add(route.name)
    paths.add(route.path)
    shapes.add(shape)
    return Object.freeze({ name: route.name, path: route.path })
  })

  return Object.freeze(entries)
}

export const pageRouteEntries = parsePageRouteManifest(pageRouteManifest)

const pathsByName = new Map(pageRouteEntries.map((route) => [route.name, route.path]))

export function pagePath(name: string): string {
  const routePath = pathsByName.get(name)
  if (routePath === undefined) {
    throw new Error(`unknown page route "${name}"`)
  }
  return routePath
}

export function pagePathMatches(name: string, rawPath: string): boolean {
  const routePath = pathsByName.get(name)
  if (routePath === undefined) return false

  let decodedPath: string
  try {
    decodedPath = decodeURIComponent(rawPath)
  } catch {
    return false
  }
  if (routePath === '/' || decodedPath === '/') {
    return routePath === decodedPath
  }

  const routeSegments = routePath.slice(1).split('/')
  const pathSegments = decodedPath.startsWith('/') ? decodedPath.slice(1).split('/') : []
  if (routeSegments.length !== pathSegments.length) return false

  return routeSegments.every((segment, index) => {
    const pathSegment = pathSegments[index]
    return parameterSegmentPattern.test(segment)
      ? pathSegment !== undefined && pathSegment !== ''
      : segment === pathSegment
  })
}

File: internal/webui/page_routes.go
package webui

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"strings"
)

const pageRouteManifestVersion = 1

var (
	//go:embed page_routes.json
	embeddedPageRouteManifest []byte

	//go:embed modern_page_routes.json
	embeddedModernPageRouteManifest []byte

	staticPageSegmentPattern    = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._~-]*$`)
	parameterPageSegmentPattern = regexp.MustCompile(`^:[A-Za-z][A-Za-z0-9_]*$`)
)

type pageRoute struct {
	Name string `json:"name"`
	Path string `json:"path"`
}

type pageRouteManifest struct {
	Version int         `json:"version"`
	Routes  []pageRoute `json:"routes"`
}

func loadPageRoutes() ([]pageRoute, error) {
	routes, err := parsePageRouteManifest(embeddedPageRouteManifest)
	if err != nil {
		return nil, err
	}
	modernRoutes, err := parsePageRouteManifest(embeddedModernPageRouteManifest)
	if err != nil {
		return nil, err
	}
	// 新版地址只在服务端合并，旧版继续读取原有页面清单。
	return validatePageRoutes(append(routes, modernRoutes...))
}

func parsePageRouteManifest(data []byte) ([]pageRoute, error) {
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()

	var manifest pageRouteManifest
	if err := decoder.Decode(&manifest); err != nil {
		return nil, fmt.Errorf("decode page route manifest: %w", err)
	}
	if err := ensurePageRouteManifestEOF(decoder); err != nil {
		return nil, err
	}
	if manifest.Version != pageRouteManifestVersion {
		return nil, fmt.Errorf(
			"page route manifest version = %d, want %d",
			manifest.Version,
			pageRouteManifestVersion,
		)
	}
	if len(manifest.Routes) == 0 {
		return nil, fmt.Errorf("page route manifest routes must not be empty")
	}
	return validatePageRoutes(manifest.Routes)
}

func validatePageRoutes(entries []pageRoute) ([]pageRoute, error) {
	names := make(map[string]struct{}, len(entries))
	paths := make(map[string]struct{}, len(entries))
	shapes := make(map[string]struct{}, len(entries))
	routes := make([]pageRoute, len(entries))
	for index, route := range entries {
		if route.Name == "" || strings.TrimSpace(route.Name) != route.Name {
			return nil, fmt.Errorf("page route at index %d has invalid name", index)
		}
		if !validPageRoutePath(route.Path) {
			return nil, fmt.Errorf(
				"page route %q has unsupported shared path %q",
				route.Name,
				route.Path,
			)
		}
		if _, exists := names[route.Name]; exists {
			return nil, fmt.Errorf("duplicate page route name %q", route.Name)
		}
		if _, exists := paths[route.Path]; exists {
			return nil, fmt.Errorf("duplicate page route path %q", route.Path)
		}
		shape := pageRouteShape(route.Path)
		if _, exists := shapes[shape]; exists {
			return nil, fmt.Errorf("duplicate page route shape %q", shape)
		}
		names[route.Name] = struct{}{}
		paths[route.Path] = struct{}{}
		shapes[shape] = struct{}{}
		routes[index] = route
	}
	return routes, nil
}

func ensurePageRouteManifestEOF(decoder *json.Decoder) error {
	var extra any
	if err := decoder.Decode(&extra); err == io.EOF {
		return nil
	} else if err != nil {
		return fmt.Errorf("decode trailing page route manifest data: %w", err)
	}
	return fmt.Errorf("page route manifest contains trailing JSON values")
}

func pageRouteShape(value string) string {
	segments := strings.Split(value, "/")
	for index, segment := range segments {
		if parameterPageSegmentPattern.MatchString(segment) {
			segments[index] = ":"
		}
	}
	return strings.Join(segments, "/")
}

func validPageRoutePath(value string) bool {
	if value == "/" {
		return true
	}
	if !strings.HasPrefix(value, "/") || strings.HasSuffix(value, "/") {
		return false
	}
	for _, segment := range strings.Split(strings.TrimPrefix(value, "/"), "/") {
		if parameterPageSegmentPattern.MatchString(segment) {
			continue
		}
		if !staticPageSegmentPattern.MatchString(segment) || segment == "." || segment == ".." {
			return false
		}
	}
	return true
}
